src/grpc/infer_handler.h

```diff
@@ -656,25 +640,86 @@ class InferHandlerState {
 
     void GrpcContextAsyncNotifyWhenDone(InferHandlerStateType* state)
     {
-      ctx_->AsyncNotifyWhenDone(state);
+      InferHandlerStateType* wrapped_state =
+          new InferHandlerStateType(Steps::WAITING_NOTIFICATION, state);
+      notify_state_ = state;
+      ctx_->AsyncNotifyWhenDone(wrapped_state);
```

Comment:
Since we already have state, wrapped_state, and wrapped_state->state_ptr, I think we should avoid having yet another state with notify_state. Let's instead toggle some boolean flag on the state like `state.is_notify_state_` or something and check that rather than comparing state pointers.

Reply:
Done!